15 digit numbers on 32 bit machiners

bitnissen opened the pull request:
On 32 bit machines, convert to a string containing the integer value by removing all digits and ensuring that the string doesn't start with a 0.

On 64 bit machines, still just cast to int, as 64 bit is long enough to handle 15 digits in base10.

bitnissen committed:
15 digit numbers on 32 bit machiners

cypres commented:
Why not just send the MSISDN as a string? It would seem less hacky?

bitnissen commented:
@duedal Indeed, but the `gwapi_send_sms`-function is not considered internal, but a method which any theme or plugin developer may use. I try to clean up the data sent as much as possible before handing it over to the gateway, so I'd like to ensure that especially the MSISDN is properly formatted.

In my opinion it would perhaps make sense to be even more strict in validating the user input in this function, so it's nearly impossible for it to send anything invalid to the API, perhaps doing a more thorough validation of the MSISDNs. This is a step in that direction, which is now also 32-bit compatible.

Do you agree to this approach?

💾

cypres commented on inc/api.php:
```diff
@@ -59,7 +59,7 @@ function gwapi_send_sms($message, $recipients, $sender='', $destaddr='MOBILE')
     foreach($recipients_formatted as $msisdn => $tags) {
         $rec = [
-            'msisdn' => (int)$msisdn,
+            'msisdn' => gwapi_bigint($msisdn),
```
Then this will do the trick.

``` php
filter_var($msisdn, FILTER_SANITIZE_NUMBER_INT)
```

bitnissen replied:
Ah, cool, didn't remember that it also did that :)

bitnissen replied:
Hm, almost... My method still beats it in this case:
`<?php echo filter_var("(+45) 1234567890 1234567890", FILTER_SANITIZE_NUMBER_INT);`
which outputs:
`+4512345678901234567890`

I don't believe the gateway supports the + character prefixed to the MSISDN?

cypres replied:
It does, and we will keep this behavior for the same reasons, although the MSISDN does not include the + prefix, this is a very common mistake so we will allow it.

bitnissen committed:
Update api.php

Re-implemented using just filter_var for sanitation for all platforms, instead of the double solution before (32v64 bit).

cypres commented:
LGTM :shipit:
